core/shared/src/main/scala/fs2/text.scala

```diff
@@ -305,12 +308,100 @@ object text {
     utf8.decodeC
 
   /** Encodes a stream of `String` in to a stream of bytes using the given charset. */
-  def encode[F[_]](charset: Charset): Pipe[F, String, Byte] =
-    _.mapChunks(c => c.flatMap(s => Chunk.array(s.getBytes(charset))))
+  def encode[F[_]](charset: Charset): Pipe[F, String, Byte] = {
+    def encodeC(
+        encoder: CharsetEncoder,
+        acc: Chunk[Char],
+        s: Stream[F, Chunk[Char]],
+        lastOutBuffer: ByteBuffer
+    ): Pull[F, Byte, Unit] =
+      s.pull.uncons1.flatMap { r =>
+        val toEncode = r match {
+          case Some((c, _)) => acc ++ c
+          case None         => acc
+        }
+
+        val isLast = r.isEmpty
+        (lastOutBuffer: Buffer).clear()
+
+        val outBufferSize = (encoder.maxBytesPerChar() * toEncode.size).ceil.toInt
+
+        val out =
+          if (outBufferSize > lastOutBuffer.remaining())
+            ByteBuffer.allocate(outBufferSize)
+          else lastOutBuffer
+
+        val inBuffer = toEncode.toCharBuffer
+        val result = encoder.encode(inBuffer, out, isLast)
+        (out: Buffer).flip()
+
+        val nextAcc =
+          if (inBuffer.remaining() > 0) Chunk.charBuffer(inBuffer.slice()) else Chunk.empty
+
+        val rest = r match {
+          case Some((_, tail)) => tail
+          case None            => Stream.empty
+        }
+
+        if (result.isError) {
+          ApplicativeThrow[Pull[F, Byte, *]].raiseError {
```

Comment:
This will throw an exception. The current behaviour seems to rather replace characters with some placeholder, but on decoding there's a `RaiseThrowable` constraint. So maybe this should actually be named `encodeWithCharset` and the current implementations deprecated. Or I need to dig into mimicking the old behaviour. Opinions?

Reply:
Sorry, can you explain more about this? The docs for `CharsetEncoder` say that

> How an encoding error is handled depends upon the action requested for that type of error, which is described by an instance of the [CodingErrorAction](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CodingErrorAction.html) class. The possible error actions are to [ignore](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CodingErrorAction.html#IGNORE) the erroneous input, [report](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CodingErrorAction.html#REPORT) the error to the invoker via the returned [CoderResult](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CoderResult.html) object, or [replace](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CodingErrorAction.html#REPLACE) the erroneous input with the current value of the replacement byte array. The replacement is initially set to the encoder's default replacement, which often (but not always) has the initial value { (byte)'?' }; its value may be changed via the [replaceWith](https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetEncoder.html#replaceWith-byte:A-) method.

https://docs.oracle.com/javase/8/docs/api/java/nio/charset/CharsetEncoder.html

Reply:
The default with `CharsetEncoder` is to raise an error, different from what `String.getBytes` does. See this little Scastie: https://scastie.scala-lang.org/2pOZGh6PQBiJ8TQyd5bYkQ

It seems totally possible to retain the old behaviour and just replace any non-mappable character, but it is somewhat inconsistent with `decode` which deals with errors by making them visible in `F`. So this PR would be a good opportunity to revisit this and potentially align both ways (which isn't straight-forward due to bin/src compat).

Reply:
Just pushed b010a950 that restores the replacing behaviour.